### PR TITLE
Support package selection options like `--exclude` in `cargo publish`

### LINKED ISSUE #14652: `cargo publish` doesn't allow `--workspace`, multiple `--package`, or `--exclude`
Was testing out #1169 in `cargo-release` where we exclusively use `--package` and found it didn't work.

### PR DESCRIPTION
**Pull request by jneem**
Fixes #14652.

Is there a way to make the help text depend on whether nightly/unstable features are enabled? I couldn't find one...

### CODE REVIEW

File: src/bin/cargo/commands/publish.rs [hunk 1]
```diff
@@ -18,7 +18,11 @@ pub fn cli() -> Command {
             "Allow dirty working directories to be packaged",
         ))
         .arg_silent_suggestion()
-        .arg_package("Package to publish")
+        .arg_package_spec_no_all(
+            "Package(s) to publish",
+            "Publish all packages in the workspace (requires nightly)",
+            "Don't publish specified packages (requires nightly)",
```
Comment (weihanglo) [inline src/bin/cargo/commands/publish.rs:24]: By convention we use a shorter `(unstable)` to denote that.

```suggestion
            "Publish all packages in the workspace (unstable)",
            "Don't publish specified packages (unstable)",
```

File: tests/testsuite/publish.rs [hunk 2]
```diff
@@ -3364,6 +3364,94 @@ You may press ctrl-c to skip waiting; the crate should be available shortly.
...
+[PACKAGING] a v0.1.0 ([ROOT]/foo/a)
+[PACKAGED] 3 files, [FILE_SIZE]B ([FILE_SIZE]B compressed)
+[UPLOADING] a v0.1.0 ([ROOT]/foo/a)
+[WARNING] aborting upload due to dry run
+
+"#]])
+        .with_stdout_data(str![[r#""#]])
+        .run();
+
+    p.cargo("publish --no-verify --dry-run --package a --package b")
```
Comment (weihanglo) [inline tests/testsuite/publish.rs:3444]: > Is there a way to make the help text depend on whether nightly/unstable features are enabled? I couldn't find one...

I am not aware of any. You may need to follow what you have done manually. We may need nightly-gate tests for `--workspace` and `exclude` for both without `+nightly` and `-Zpackage-workspace` situations.

https://github.com/rust-lang/cargo/blob/15fbd2f607d4defc87053b8b76bf5038f2483cf4/src/bin/cargo/commands/package.rs#L55-L62

File: tests/testsuite/publish.rs
Comment (weihanglo) [inline tests/testsuite/publish.rs:1]: Should we have [`cargo publish` manpage](https://github.com/rust-lang/cargo/blob/15fbd2f607d4defc87053b8b76bf5038f2483cf4/src/doc/man/cargo-publish.md#L71) updated to include `--workspace` and `--exclude`?

Comment (jneem) [reply tests/testsuite/publish.rs:1]: I'd be happy to update it if you want. Is there a policy for unstable features in manpages?

Comment (weihanglo) [reply tests/testsuite/publish.rs:1]: Thank you. We already did that for a couple of flags. See some examples:

* https://doc.rust-lang.org/1.81.0/cargo/commands/cargo-build.html#option-cargo-build---artifact-dir
* https://doc.rust-lang.org/1.81.0/cargo/commands/cargo-build.html#option-cargo-build---build-plan
* https://doc.rust-lang.org/1.81.0/cargo/commands/cargo-build.html#option-cargo-build--C

I think generally adding a sentence noting the unstable status is good enough.

**Comment by weihanglo** [comment]
I presume that these flags are behind `-Zpackage-workspace` so we don't need to go through the RFC process until the stabilization. Once the man pages are added I'll merge this.

**Review by weihanglo — APPROVED** [review]
Thanks!

We'll need to clean up duplicate man page (d30fde980f524f0240aa5ccb1a10199aff5f5de1) when stabilization.

**Pull request merged**